File: Go/MonteCarloSimulationPi/MonteCarloSimulationPi.go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

// mcs stands for Monte Carlo Simulation
// pi=3.1415926535

// MonteCarloSimulationPi is a structure to calculate pi using Monte Carlo simulation.
type MonteCarloSimulationPi struct {
	numberOfProcesses int
	parallelFlag      bool
	experimentFlag    bool
}

func (monteCarloSimulationPi *MonteCarloSimulationPi) simulationPi(numberOfSimulations int, channel chan int) {
	if monteCarloSimulationPi.experimentFlag == true {
		inside := 0
		s := rand.NewSource(time.Now().UnixNano())
		r := rand.New(s)
		for i := 0; i < numberOfSimulations; i++ {
			x := r.Float64()
			y := r.Float64()
			if (x*x + y*y) < 1 {
				inside++
			}
		}
		channel <- inside
	} else {
		//One row in the output file.
		var row strings.Builder
		inside := 0
		// Source code for random number generator https://play.golang.org/p/ZdFpbahgC1
		// The default number generator is deterministic, so it'll
		// produce the same sequence of numbers each time by default.
		// To produce varying sequences, give it a seed that changes.
		// Note that this is not safe to use for random numbers you
		// intend to be secret, use `crypto/rand` for those.
		// Seeding - Go provides a method, Seed(seed int64), that allows you
		// to initialize this default sequence. Implementation is slow
		// to make it faster rand.Seed(time.Now().UnixNano()) is added.
		// Seed is the current time, converted to int64 by UnixNano.
		// Gives constantly changing numbers.

		// Seed
		s := rand.NewSource(time.Now().UnixNano())
		// Randomly changing numbers.
		r := rand.New(s)
		for i := 0; i < numberOfSimulations; i++ {
			// Call the resulting `rand.Rand` just like the
			// functions on the `rand` package.
			x := r.Float64()
			y := r.Float64()

			// Pharo for Data Visualization. Circle of radius 250 centered at the point(250, 250).
			// To create a Rectangle in Pharo you must provide the top left and the bottom right points.
			row.WriteString(strconv.FormatInt(int64(x*500), 10) + " " + strconv.FormatInt(int64(y*500), 10) + "\r\n")
			if (x*x + y*y) < 1 {
				inside++
			}
		}
		monteCarloSimulationPi.exportPiFile(row.String())
		channel <- inside
	}

}

func (monteCarloSimulationPi *MonteCarloSimulationPi) mcsPiSerial(numberOfSimulations int) (float64, float64) {
	startTime := time.Now()
	monteCarloSimulationPi.parallelFlag = false
	channel := make(chan int)
	go monteCarloSimulationPi.simulationPi(numberOfSimulations, channel)
	inside := <-channel
	pi := 4 * float64(inside) / float64(numberOfSimulations)
	executionTime := time.Since(startTime).Seconds()
	return pi, executionTime
}

func (monteCarloSimulationPi *MonteCarloSimulationPi) mcsPiParallel(numberOfSimulations int) (float64, float64) {
	startTime := time.Now()
	monteCarloSimulationPi.parallelFlag = true
	numberOfSimulationsPerProcess := numberOfSimulations / monteCarloSimulationPi.numberOfProcesses
	/* 	Buffered channels are useful when you know how many goroutines you have launched,
	   	want to limit the number of goroutines you will launch, or want to limit
	   	the amount of work that is queued up. */
	channel := make(chan int, monteCarloSimulationPi.numberOfProcesses)

	for i := 0; i < monteCarloSimulationPi.numberOfProcesses; i++ {
		go monteCarloSimulationPi.simulationPi(numberOfSimulationsPerProcess, channel)
	}

	var inside int
	for i := 0; i < monteCarloSimulationPi.numberOfProcesses; i++ {
		inside += <-channel
	}
	pi := 4 * float64(inside) / float64(numberOfSimulations)
	executionTime := time.Since(startTime).Seconds()
	return pi, executionTime

}

func (monteCarloSimulationPi *MonteCarloSimulationPi) exportPiFile(simulations string) {
	var path string
	if monteCarloSimulationPi.parallelFlag == false {
		path = "C:\\Users\\Dule\\Desktop\\NAPREDNE TEHNIKE PROGRAMIRANJA\\PROJEKAT\\NTP\\Execution Results\\Pi\\GolangPiSerial.txt"
	} else {
		path = "C:\\Users\\Dule\\Desktop\\NAPREDNE TEHNIKE PROGRAMIRANJA\\PROJEKAT\\NTP\\Execution Results\\Pi\\GolangPiParallel.txt"
	}
	f, err := os.Create(path) // creating...
	if err != nil {
		fmt.Printf("Error while creating a file: %v", err)
		return
	}
	defer f.Close()
	// Writing to file
	_, err = f.WriteString(simulations)
	if err != nil {
		fmt.Printf("Error while writing a file: %v", err)
	}

}

func exportScalingFile(scaling string, strongFlag bool) {
	var path string
	if strongFlag == true {
		path = "C:\\Users\\Dule\\Desktop\\NAPREDNE TEHNIKE PROGRAMIRANJA\\PROJEKAT\\NTP\\Scaling Results\\Pi\\GolangPiStrongScaling.csv"
	} else {
		path = "C:\\Users\\Dule\\Desktop\\NAPREDNE TEHNIKE PROGRAMIRANJA\\PROJEKAT\\NTP\\Scaling Results\\Pi\\GolangPiWeakScaling.csv"
	}
	f, err := os.Create(path) // creating...
	if err != nil {
		fmt.Printf("Error while creating a file: %v", err)
		return
	}
	defer f.Close()
	// Writing to file
	_, err = f.WriteString(scaling)
	if err != nil {
		fmt.Printf("Error while writing a file: %v", err)
	}

}

var serialPart float64 = 0.0
var parallelPart float64 = 1.0

func calculateAmdahlSpeedup(numberOfProcesses int) float64 {
	return 1.0 / (serialPart + parallelPart/float64(numberOfProcesses))
}

func calculateGustafsonSpeedup(numberOfProcesses int) float64 {
	return serialPart + parallelPart*float64(numberOfProcesses)
}

func strongScaling() {
	fmt.Println("=======================")
	fmt.Println("Start strong scaling:")
	fmt.Println("=======================")
	//One row in the output file.
	var row strings.Builder
	row.WriteString("number_of_processes,achieved_speedup,theoretical_maximum_speedup\r\n")
	numberOfSimulations := 100000000
	numberOfProcessesSerial := 1
	monteCarloSimulationPiSerial := MonteCarloSimulationPi{numberOfProcesses: numberOfProcessesSerial}
	monteCarloSimulationPiSerial.experimentFlag = true
	fmt.Println("Approximation of Pi by using the Monte Carlo simulation serial version")
	serialPi, serialExecutionTime := monteCarloSimulationPiSerial.mcsPiSerial(numberOfSimulations)
	fmt.Printf("Pi(n = %d, p = %d) = %f\r\n", numberOfSimulations, numberOfProcessesSerial, serialPi)
	fmt.Printf("Execution time (duration): %f seconds\r\n", serialExecutionTime)
	for numberOfProcessesParallel := 2; numberOfProcessesParallel < 14; numberOfProcessesParallel++ {
		monteCarloSimulationPiParallel := MonteCarloSimulationPi{numberOfProcesses: numberOfProcessesParallel}
		monteCarloSimulationPiParallel.experimentFlag = true
		fmt.Println("Approximation of Pi by using the Monte Carlo simulation parallel version")
		parallelPi, parallelExecutionTime := monteCarloSimulationPiParallel.mcsPiParallel(numberOfSimulations)
		fmt.Printf("Pi(n = %d, p = %d) = %f\r\n", numberOfSimulations, numberOfProcessesParallel, parallelPi)
		fmt.Printf("Execution time (duration): %f seconds.\r\n", parallelExecutionTime)
		achievedSpeedup := serialExecutionTime / parallelExecutionTime
		theoreticalMaximumSpeedup := calculateAmdahlSpeedup(numberOfProcessesParallel)
		fmt.Printf("Achieved speedup is: %f times.\r\n", achievedSpeedup)
		fmt.Printf("Theoretical maximum speedup according to Amdahl’s law is: %f times.\r\n", theoreticalMaximumSpeedup)
		row.WriteString(strconv.FormatInt(int64(numberOfProcessesParallel), 10) + "," + strconv.FormatFloat(achievedSpeedup, 'f', -1, 64) + "," + strconv.FormatInt(int64(theoreticalMaximumSpeedup), 10) + "\r\n")
	}
	exportScalingFile(row.String(), true)
	fmt.Println("End strong scaling.")

}

func weakScaling() {
	fmt.Println("=======================")
	fmt.Println("Start weak scaling:")
	fmt.Println("=======================")
	//One row in the output file.
	var row strings.Builder
	row.WriteString("number_of_processes,achieved_speedup,theoretical_maximum_speedup\r\n")
	numberOfSimulations := 100000000
	for numberOfProcesses := 2; numberOfProcesses < 14; numberOfProcesses++ {
		increasedNumberOfSimulations := numberOfSimulations * numberOfProcesses
		monteCarloSimulationPi := MonteCarloSimulationPi{numberOfProcesses: numberOfProcesses}
		monteCarloSimulationPi.experimentFlag = true
		fmt.Println("Approximation of Pi by using the Monte Carlo simulation serial version")
		serialPi, serialExecutionTime := monteCarloSimulationPi.mcsPiSerial(increasedNumberOfSimulations)
		fmt.Printf("Pi(n = %d, p = %d) = %f\r\n", increasedNumberOfSimulations, 1, serialPi)
		fmt.Printf("Execution time (duration): %f seconds\r\n", serialExecutionTime)
		fmt.Println("Approximation of Pi by using the Monte Carlo simulation parallel version")
		parallelPi, parallelExecutionTime := monteCarloSimulationPi.mcsPiParallel(increasedNumberOfSimulations)
		fmt.Printf("Pi(n = %d, p = %d) = %f\r\n", increasedNumberOfSimulations, numberOfProcesses, parallelPi)
		fmt.Printf("Execution time (duration): %f seconds.\r\n", parallelExecutionTime)
		achievedSpeedup := serialExecutionTime / parallelExecutionTime
		theoreticalMaximumSpeedup := calculateAmdahlSpeedup(numberOfProcesses)
		fmt.Printf("Achieved speedup is: %f times.\r\n", achievedSpeedup)
		fmt.Printf("Theoretical maximum speedup according to Amdahl’s law is: %f times.\r\n", theoreticalMaximumSpeedup)
		row.WriteString(strconv.FormatInt(int64(numberOfProcesses), 10) + "," + strconv.FormatFloat(achievedSpeedup, 'f', -1, 64) + "," + strconv.FormatInt(int64(theoreticalMaximumSpeedup), 10) + "\r\n")
	}
	exportScalingFile(row.String(), false)
	fmt.Println("End weak scaling.")

}

func main() {
	numberOfSimulationsSerial := 1000
	numberOfProcessesSerial := 1
	monteCarloSimulationPiSerial := MonteCarloSimulationPi{numberOfProcesses: numberOfProcessesSerial}
	monteCarloSimulationPiSerial.experimentFlag = false
	fmt.Println("Approximation of Pi by using the Monte Carlo simulation serial version")
	serialPi, serialExecutionTime := monteCarloSimulationPiSerial.mcsPiSerial(numberOfSimulationsSerial)
	fmt.Printf("Pi(n = %d, p = %d) = %f\r\n", numberOfSimulationsSerial, numberOfProcessesSerial, serialPi)
	fmt.Printf("Execution time (duration): %f seconds\r\n", serialExecutionTime)

	numberOfSimulationsParallel := 4000
	numberOfProcessesParallel := 4
	monteCarloSimulationPiParallel := MonteCarloSimulationPi{numberOfProcesses: numberOfProcessesParallel}
	monteCarloSimulationPiParallel.experimentFlag = false
	fmt.Println("Approximation of Pi by using the Monte Carlo simulation parallel version")
	parallelPi, parallelExecutionTime := monteCarloSimulationPiParallel.mcsPiParallel(numberOfSimulationsParallel)
	fmt.Printf("Pi(n = %d, p = %d) = %f\r\n", numberOfSimulationsParallel, numberOfProcessesParallel, parallelPi)
	fmt.Printf("Execution time (duration): %f seconds", parallelExecutionTime)

	//strongScaling()
	//weakScaling()

}

File: Python/MonteCarloSimulationPi/MonteCarloSimulationPi.py
import random
import time
from multiprocessing import Pool


def calculate_execution_time(function):
    def calculate_duration(*args, **kwargs):
        start_time = time.time()
        executing_function = function(*args, **kwargs)
        end_time = time.time()
        execution_time = round(end_time - start_time, 7)
        return executing_function, execution_time
    return calculate_duration


class MonteCarloSimulationPi:
    def __init__(self, number_of_processes):
        self.number_of_processes = number_of_processes
        self.parallel_flag = False
        self.experiment_flag = False

    # mcs stands for Monte Carlo Simulation
    # pi=3.1415926535
    def simulation_pi(self, number_of_simulations):
        if self.experiment_flag == True:
            inside = 0
            for _ in range(number_of_simulations):
                x = random.random()
                y = random.random()
                # The unit circle is the circle of radius 1 centered at the origin(0, 0)
                # in the Cartesia coordinate system in the Euclidean plane.
                if x * x + y * y < 1:
                    inside = inside + 1
            return inside
        else:
            if self.parallel_flag == False:
                # r before string converts normal string to raw string
                path = r"C:\Users\Dule\Desktop\NAPREDNE TEHNIKE PROGRAMIRANJA\PROJEKAT\NTP" \
                       r"\Execution Results\Pi\PythonPiSerial.txt"
            else:
                # r before string converts normal string to raw string
                path = r"C:\Users\Dule\Desktop\NAPREDNE TEHNIKE PROGRAMIRANJA\PROJEKAT\NTP" \
                       r"\Execution Results\Pi\PythonPiParallel.txt"
            out_file = open(path, "w")
            inside = 0
            for _ in range(number_of_simulations):
                x = random.random()
                y = random.random()
                # Pharo for Data Visualization. Circle of radius 250 centered at the point(250, 250).
                # To create a Rectangle in Pharo you must provide the top left and the bottom right points.
                out_file.write(str(int(x * 500)) + ' ' + str(int(y * 500)) + '\n')
                # The unit circle is the circle of radius 1 centered at the origin(0, 0)
                # in the Cartesia coordinate system in the Euclidean plane.
                if x * x + y * y < 1:
                    inside = inside + 1
            out_file.close()
            return inside

    @calculate_execution_time
    def mcs_pi_serial(self, number_of_simulations):
        self.parallel_flag = False
        pi = 4 * self.simulation_pi(number_of_simulations) / number_of_simulations
        return pi

    @calculate_execution_time
    def mcs_pi_parallel(self, number_of_simulations):
        self.parallel_flag = True
        pool = Pool(processes=self.number_of_processes)
        number_of_simulations_per_process = int(number_of_simulations / self.number_of_processes)
        simulations_per_process = []
        # Append the same value multiple times to a list
        # To add v, n times, to l:
        # l += n * [v]
        simulations_per_process += self.number_of_processes * [number_of_simulations_per_process]
        inside_sum = pool.map(self.simulation_pi, simulations_per_process)
        pi = 4 * sum(inside_sum) / number_of_simulations
        return pi


if __name__ == "__main__":
    number_of_simulations_serial = 1000
    number_of_processes_serial = 1
    monte_carlo_simulation_pi_serial = MonteCarloSimulationPi(number_of_processes_serial)
    monte_carlo_simulation_pi_serial.experiment_flag = False
    print("Approximation of Pi by using the Monte Carlo simulation serial version")
    serial_pi, serial_execution_time = monte_carlo_simulation_pi_serial.mcs_pi_serial(number_of_simulations_serial)
    print("Pi(n = {}, p = {}) = {}".format(number_of_simulations_serial,number_of_processes_serial,serial_pi))
    print("Execution time (duration): {} seconds".format(serial_execution_time))

    number_of_simulations_parallel = 1000
    number_of_processes_parallel = 4
    monte_carlo_simulation_pi_parallel = MonteCarloSimulationPi(number_of_processes_parallel)
    monte_carlo_simulation_pi_parallel.experiment_flag = False
    print("Approximation of Pi by using the Monte Carlo simulation parallel version")
    parallel_pi, parallel_execution_time = monte_carlo_simulation_pi_parallel.mcs_pi_parallel(number_of_simulations_parallel)
    print("Pi(n = {}, p = {}) = {}".format(number_of_simulations_parallel,number_of_processes_parallel,parallel_pi))
    print("Execution time (duration): {} seconds".format(parallel_execution_time))

